lib/lodqa/bs_client.rb

```diff
@@ -55,8 +56,71 @@ def subscribe ws, request_id, url, mode
         end
       end
 
+      def sparqls pgp, mappings, graph_finder, enum
+        anchored_pgps(pgp, mappings).each do |anchored_pgp|
+          to_sparql(anchored_pgp, graph_finder){ |sparql| enum << sparql}
+        end
+      end
+
       private
 
+      def anchored_pgps pgp, mappings
+        Logger::Logger.debug "start #{self.class.name}##{__method__}"
+
+        Enumerator.new do |anchored_pgps|
+          pgp[:nodes].delete_if{|n| nodes_to_delete(pgp, mappings).include? n}
+          pgp[:edges].uniq!
+          terms = pgp[:nodes].values.map{|n| mappings[n[:text].to_sym]}
+
+          terms.map!{|t| t.nil? ? [] : t}
+
+          Logger::Logger.debug "terms: #{ terms.first.product(*terms.drop(1)) }"
+
+          terms.first.product(*terms.drop(1))
+            .each do |ts|
+              anchored_pgp = pgp.dup
+              anchored_pgp[:nodes] = pgp[:nodes].dup
+              anchored_pgp[:nodes].each_key{|k| anchored_pgp[:nodes][k] = pgp[:nodes][k].dup}
+              anchored_pgp[:nodes].each_value.with_index{|n, i| n[:term] = ts[i]}
+
+              anchored_pgps << anchored_pgp
+            end
+        end
+      end
```

Comment:
EnumeratorはSPARQLを生成たびに実行するのに使っていました。
件数を数えるためであれば必要ありません。

単にmapを使って配列を返せば十分です。
例えば次です...
```suggestion
          anchored_pgps = []
          pgp[:nodes].delete_if{|n| nodes_to_delete(pgp, mappings).include? n}
          pgp[:edges].uniq!
          terms = pgp[:nodes].values.map{|n| mappings[n[:text].to_sym]}

          terms.map!{|t| t.nil? ? [] : t}

          Logger::Logger.debug "terms: #{ terms.first.product(*terms.drop(1)) }"

          terms.first.product(*terms.drop(1))
            .each do |ts|
              anchored_pgp = pgp.dup
              anchored_pgp[:nodes] = pgp[:nodes].dup
              anchored_pgp[:nodes].each_key{|k| anchored_pgp[:nodes][k] = pgp[:nodes][k].dup}
              anchored_pgp[:nodes].each_value.with_index{|n, i| n[:term] = ts[i]}

              anchored_pgps << anchored_pgp
            end
          end
 
          anchored_pgps
```

Reply:
指摘通り対応しました。